File: 1833. Maximum Ice Cream Bars/Solution.py
class Solution:
    def maxIceCream(self, costs: List[int], coins: int) -> int:
        # return self.pointers(costs, coins)
        # return self.prefixSum(costs, coins)
        return self.bruteForce(costs, coins)
        
    def pointers(self, costs, coins):
        freq = [0 for i in range(100001)]
        for c in costs: freq[c] += 1
        count = 0
        for i in range(100001):
            if freq[i] == 0: continue
            currMax = min(coins // i, freq[i])
            count += currMax
            coins -= currMax * i
            if (coins <= 0) and (freq[i] - currMax >= 0): break
        return count

    def prefixSum(self, costs, coins):
        costs.sort()
        count, n = 0, len(costs)
        prefix = [0 for i in range(n)]
        for i in range(n):
            prefix[i] = costs[i] if i == 0 else prefix[i - 1] + costs[i]
            if coins < prefix[i]: return i
        return n

    def bruteForce(self, costs, coins):
        costs.sort()
        n = len(costs)
        for i in range(n):
            coins -= costs[i]
            if coins < 0: return i
        return n

File: 1833. Maximum Ice Cream Bars/Solution.java
class Solution {
    public int maxIceCream(int[] costs, int coins) {
        // return prefixSumSolution(costs, coins);
        return pointersSolution(costs, coins);
    }

    private int prefixSumSolution(int[] costs, int coins) {
        Arrays.sort(costs);
        int n = costs.length, count = 0;
        int[] prefix = new int[n];
        for (int i = 0; i < n; i++) {
            prefix[i] = (i == 0) ? costs[i] : prefix[i - 1] + costs[i];
            if (prefix[i] > coins) return i;
        }
        return n;
    }

    private int pointersSolution(int[] costs, int coins) {
        int[] freq = new int[100001];
        for (int c : costs) freq[c]++;
        int count = 0;
        for (int i = 0; i < 100001; i++) {
            if (freq[i] == 0) continue;
            int currMax = Math.min(coins / i, freq[i]);
            count += currMax;
            coins -= i * currMax;
            if (coins <= 0 && freq[i] - currMax >= 0) break;
        }
        return count;

    }
}